crates/red_knot_workspace/src/project/combine.rs

```diff
@@ -0,0 +1,155 @@
+use std::{collections::HashMap, hash::BuildHasher};
+
+use crate::project::options::{EnvironmentOptions, SrcOptions};
+use red_knot_python_semantic::{PythonPath, PythonPlatform, PythonVersion};
+use ruff_db::system::SystemPathBuf;
+
+/// Combine two values, preferring the values in `self`.
+///
+/// The logic should follow that of Cargo's `config.toml`:
+///
+/// > If a key is specified in multiple config files, the values will get merged together.
+/// > Numbers, strings, and booleans will use the value in the deeper config directory taking
+/// > precedence over ancestor directories, where the home directory is the lowest priority.
+/// > Arrays will be joined together with higher precedence items being placed later in the
+/// > merged array.
+///
+/// Note: The merging behavior differs from uv in that values with higher precedence in arrays
+/// are placed later in the merged array. This is because we want to support overriding previously
+/// provided values, including unsetting them. For example: patterns coming last in file inclusion and exclusion patterns
+/// allow overriding earlier patterns, matching the `gitignore` behavior. This also closer matches
+/// a CLI interface where values provided later can override values provided earlier: `knot --ignore=correctness --warn=correctness/rule`
+/// should enable `correctness/rule` even though `correctness` was previously ignored.
+///
+/// ## Macro
+/// You can automatically derive `Combine` for structs with named fields by using `derive(ruff_macros::Combine)`.
+pub trait Combine {
```

Comment:
Minor naming comment: "combine" has no clear direction/order/precedence for me. Something like `OverrideWith`, `OverlayWith`, `ExtendWith` or similar might be a bit clearer in terms of which side is taking precedence?

Reply:
Also: I want that trait in a separate crate so I can use it in my projects! :star_struck:

Reply:
I used this name because it's what Ruff and uv are using... I considered naming it `Merge` but it didn't seem an improvement enough to justify diverging.

Reply:
Ah, I didn't realize it was an existing concept.